### PR TITLE
Voicea events

### PR DESCRIPTION
**Pull request by srinijay98**
# COMPLETES #< INSERT LINK TO ISSUE >

## This pull request addresses

< DESCRIBE THE CONTEXT OF THE ISSUE >
Do proper disconnection and release events for voicea
## by making the following changes

< DESCRIBE YOUR CHANGES >

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [ ] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request

- [ ] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/internal-plugin-llm/src/llm.ts [hunk 1]
```diff
@@ -75,53 +73,62 @@ export default class LLMChannel extends (Mercury as any) implements ILLMChannel
...
    */
-  public disconnectLLM = (): Promise<void> => this.disconnect();
+  public disconnectLLM = (): Promise<void> =>
+    this.disconnect().then(() => {
+      this.registered = false;
+      this.locusUrl = undefined;
+      this.binding = undefined;
+      this.webSocketUrl = undefined;
+    });
 }
```
Comment (srinijay98) [inline packages/@webex/internal-plugin-llm/src/llm.ts:134]: changed made here

File: packages/@webex/internal-plugin-llm/test/unit/spec/llm.js [hunk 2]
```diff
@@ -90,6 +89,9 @@ describe('plugin-llm', () => {
       it('disconnects mercury', async () => {
         await llmService.disconnect();
         sinon.assert.calledOnce(llmService.disconnect);
+        assert.equal(llmService.isConnected(), false);
+        assert.equal(llmService.getLocusUrl(), undefined);
+        assert.equal(llmService.getBinding(), undefined);
```
Comment (srinijay98) [inline packages/@webex/internal-plugin-llm/test/unit/spec/llm.js:94]: changed made here

File: packages/@webex/internal-plugin-voicea/src/voicea.ts [hunk 3]
```diff
@@ -19,32 +24,41 @@ export class VoiceaChannel extends WebexPlugin implements IVoiceaChannel {
...
-          break;
-        case VOICEA_RELAY_TYPES.TRANSLATION_RESPONSE:
-          this.processCaptionLanguageResponse(e.data.voiceaPayload);
-          break;
-        case VOICEA_RELAY_TYPES.TRANSCRIPTION:
-          this.processTranscription(e.data.voiceaPayload);
-          break;
-        default:
-          break;
-      }
```
Comment (srinijay98) [inline packages/@webex/internal-plugin-voicea/src/voicea.ts:46]: separated into different function

File: packages/@webex/internal-plugin-voicea/test/unit/spec/voicea.js [hunk 4]
```diff
@@ -67,11 +66,20 @@ describe('plugin-voicea', () => {
             eventType: 'relay.event',
             relayType: 'client.annc',
           },
-          trackingId: sinon.match.string
+          trackingId: sinon.match.string,
         });
       });
     });
 
+    describe('#deregisterEvents', () => {
```
Comment (srinijay98) [inline packages/@webex/internal-plugin-voicea/test/unit/spec/voicea.js:74]: tests for deregister

File: packages/@webex/internal-plugin-voicea/src/voicea.types.ts [hunk 5]
```diff
@@ -69,14 +70,23 @@ interface TranscriptionResponse {
...
+  message: string;
 }
 
 interface IVoiceaChannel {
   setSpokenLanguage: (languageCode: string) => Promise<void>;
   requestLanguage: (languageCode: string) => void;
   turnOnCaptions: () => undefined | Promise<void>;
-  toggleTranscribing: (activate:boolean) => undefined|Promise<void>;
+  toggleTranscribing: (activate: boolean) => undefined | Promise<void>;
+  deregisterEvents: () => undefined | Promise<void>;
```
Comment (srinijay98) [inline packages/@webex/internal-plugin-voicea/src/voicea.types.ts:82]: added public facing deregisterEvents

File: packages/@webex/internal-plugin-voicea/src/voicea.ts [hunk 6]
```diff
@@ -19,32 +24,41 @@ export class VoiceaChannel extends WebexPlugin implements IVoiceaChannel {
...
-    });
+    this.webex.internal.llm.on('event:relay.event', this.eventProcessor);
```
Comment (marcin-bazyl) [inline packages/@webex/internal-plugin-voicea/src/voicea.ts:53]: don't you need to `bind(this)`?

File: packages/@webex/internal-plugin-voicea/test/unit/spec/voicea.js [hunk 7]
```diff
@@ -67,11 +66,20 @@ describe('plugin-voicea', () => {
...
+      it('deregisters voicea service', () => {
+        voiceaService.deregisterEvents();
+        assert.equal(voiceaService.hasVoiceaJoined, false);
```
Comment (marcin-bazyl) [inline packages/@webex/internal-plugin-voicea/test/unit/spec/voicea.js:77]: are these values set to anything different before the call to `voiceaService.deregisterEvents();` is done? I don't see that

File: packages/@webex/internal-plugin-llm/src/llm.ts [hunk 8]
```diff
@@ -75,53 +73,62 @@ export default class LLMChannel extends (Mercury as any) implements ILLMChannel
...
 
   /**
    * Disconnects websocket connection
    * @returns {Promise<void>}
    */
-  public disconnectLLM = (): Promise<void> => this.disconnect();
+  public disconnectLLM = (): Promise<void> =>
+    this.disconnect().then(() => {
+      this.registered = false;
+      this.locusUrl = undefined;
```
Comment (marcin-bazyl) [inline packages/@webex/internal-plugin-llm/src/llm.ts:130]: the type for `this.locusUrl` is defined as `string` so aren't you getting eslint warnings here? shouldn't the type be defined as `private locusUrl?: string;`
same applies to other fields being reset here

File: packages/@webex/internal-plugin-voicea/src/voicea.ts [hunk 9]
```diff
@@ -19,32 +24,41 @@ export class VoiceaChannel extends WebexPlugin implements IVoiceaChannel {
...
+  }
+
+  public deregisterEvents() {
+    this.hasVoiceaJoined = false;
+    this.areCaptionsEnabled = false;
+    this.isTranscribingEnabled = false;
+    this.vmcDeviceId = undefined;
```
Comment (marcin-bazyl) [inline packages/@webex/internal-plugin-voicea/src/voicea.ts:60]: same as other comment, `vmcDeviceId` type is defined as a string, so undefined is not allowed

**Comment by srinijay98** [comment]
My bad, pushed to webex repo instead of fork, will delete branch after merge

File: packages/@webex/internal-plugin-voicea/src/voicea.ts (hunk 9 above)
Comment (srinijay98) [reply packages/@webex/internal-plugin-voicea/src/voicea.ts:61]: modified

File: packages/@webex/internal-plugin-llm/src/llm.ts (hunk 8 above)
Comment (srinijay98) [reply packages/@webex/internal-plugin-llm/src/llm.ts:130]: modified

File: packages/@webex/internal-plugin-voicea/test/unit/spec/voicea.js (hunk 7 above)
Comment (srinijay98) [reply packages/@webex/internal-plugin-voicea/test/unit/spec/voicea.js:97]: my bad, added now

File: packages/@webex/internal-plugin-voicea/src/voicea.ts (hunk 6 above)
Comment (marcin-bazyl) [reply packages/@webex/internal-plugin-voicea/src/voicea.ts:54]: ah, yeah I didn't notice that you defined `this.eventProcessor` with an arrow function, so yeah that's why it works.

**Review by marcin-bazyl — APPROVED** [review]

**Pull request merged**